In [2]:
print('hello')

hello


In [3]:
data = """
Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves observing how machines work and is deeply fascinated by artificial intelligence and its impact on the IT industry.

Deepak has strong technical skills in Python, Django, Flask, and other Python tools, along with knowledge of HTML, CSS, and JavaScript. He has worked on multiple projects including a Netflix homepage clone using HTML and CSS, an e-commerce website clone, and a React calculator app that performs basic math operations with responsive design. He also built a GitHub repository called 'Kumar Projects' where he uploads data science-related work, such as the 'Analysis and Visualization of Global Population Data in 2021'. His dataset work includes analyzing data with columns such as Country Name and Population. Deepak’s GitHub projects are organized from basic to advanced, showcasing his growth in data science.

He is highly interested in statistics and hypothesis testing, and he has completed day 6 of a 'Statistics for Data Science' course. He often requests clear, step-by-step explanations in simple words and prefers visual aids or diagrams to fully understand complex topics. He explores advanced Python concepts such as interning, object identity for mutable and immutable types, duck typing, private member access, and the use of super() in class hierarchies. He also practices embeddings, vector search, and cosine similarity for NLP, where he once faced a 'ModuleNotFoundError: No module named sentence_transformers' issue and learned to install the missing library. Deepak relates the concept of vector norms to the Pythagoras theorem for better understanding.

Deepak is also active on LinkedIn, where he shares knowledge through carousels and posts about hidden Python concepts, tips, and memory optimization tricks. He once created a carousel explaining single-element tuples, another on Python interning, and one comparing object identity between immutable and mutable types. He prefers using dark, professional backgrounds for his carousel designs, with faded tones to highlight the content. His LinkedIn goals include growing his presence, creating a group focused on overlooked Python concepts, optimizing content with effective hashtags, and delivering value to the Python community.

In addition to data science, Deepak works on voice assistant projects. One of his key projects is JARVIS, a desktop voice assistant integrated with OpenAI. He experiments with libraries like pyttsx3, speech_recognition, and win32com for text-to-speech, and explores building user interfaces using Tkinter. He is curious about combining speech recognition with text-to-speech seamlessly on Windows. He enjoys hands-on experimentation and is enthusiastic about building real-world AI-powered applications.

Deepak follows Vishwa Mohan on YouTube for inspiration and learning. His personality blends technical curiosity with dedication and creativity. He enjoys exploring new concepts, sharing insights with others, and documenting his journey in tech. He values clarity and precision in communication and prefers avoiding pronouns in datasets to keep information unambiguous. Overall, Deepak is a motivated and passionate learner with strong technical skills, a growing project portfolio, and a clear vision to become a successful Data Scientist.
"""


In [4]:
clean_data = data.strip()

In [6]:
print(clean_data)

Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves observing how machines work and is deeply fascinated by artificial intelligence and its impact on the IT industry.

Deepak has strong technical skills in Python, Django, Flask, and other Python tools, along with knowledge of HTML, CSS, and JavaScript. He has worked on multiple projects including a Netflix homepage clone using HTML and CSS, an e-commerce website clone, and a React calculator app that performs basic math operations with responsive design. He also built a GitHub repository called 

In [7]:
len(clean_data)

3768

In [11]:
max_char = 500
overlap = 100
chunks = []
i = 0
while i< len(clean_data):
    piece = clean_data[i:i+max_char]
    chunks.append(piece)
    i = i+ max_char - overlap

In [12]:
chunks

['Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves',
 'himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves observing how machines work and is deeply fascinated by artificial intelligence and its impact on the IT industry.\n\nDeepak has strong technical skills in Python, Django, Flask, and other Python tools, along with knowledge of HTML, CSS, and JavaScript. He has worked on multiple projects including a Netflix homepage clone using HTML and CSS, an e-commerce website clone, and a React calcula

In [13]:
len(chunks)

10

In [14]:
import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        
        "Content-Type": "application/json",
        "Authorization": "Bearer euri-4aca7f0066db9e4df1204de08ff959f1dd8c0279d8750eff36de3a7bdb1e6101"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

In [15]:
for i in chunks:
    embedding = generate_embeddings(i)
    print(embedding)

[ 0.02237804  0.01795307 -0.00452745 ...  0.01837507  0.03518271
 -0.02058153]
[-0.01347865 -0.0321925   0.00223342 ... -0.01958637  0.00830207
 -0.0176069 ]
[-0.03493746 -0.004696   -0.02063557 ...  0.02096758  0.01015817
 -0.03258786]
[-0.00212678 -0.00915796  0.06800397 ... -0.00545227  0.03020333
 -0.03673808]
[-0.00049984  0.00327688  0.01935099 ... -0.00886498  0.01717274
 -0.02902649]
[ 0.00783338 -0.02858108 -0.00320946 ...  0.00291059  0.02553669
 -0.01619502]
[-0.00478342 -0.0259156   0.05941101 ... -0.00689318  0.00710032
 -0.02335319]
[-0.01993684 -0.02076978  0.01989654 ...  0.01121111  0.01378383
 -0.00910189]
[ 0.04160818 -0.02024986 -0.00312095 ...  0.01302452  0.023751
  0.00761059]
[ 0.0346763  -0.0179922   0.01305812 ...  0.01666115  0.02551955
 -0.00944935]


In [16]:
%pip install faiss-cpu

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\kumar\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [17]:
%pip install numpy requests

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\kumar\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [20]:
embed_list = []
meta = []

for idx , chunk in enumerate(chunks):
    vec = generate_embeddings(chunk)
    embed_list.append(vec.astype('float32'))
    meta.append({"id":idx , "text":chunk})

In [21]:
embed_list

[array([ 0.02237804,  0.01795307, -0.00452745, ...,  0.01837507,
         0.03518271, -0.02058153], shape=(1536,), dtype=float32),
 array([-0.01348186, -0.0321741 ,  0.00224046, ..., -0.01961708,
         0.00827799, -0.01762411], shape=(1536,), dtype=float32),
 array([-0.03488477, -0.00471493, -0.02062185, ...,  0.02100492,
         0.0101577 , -0.03258635], shape=(1536,), dtype=float32),
 array([-0.00212678, -0.00915796,  0.06800397, ..., -0.00545227,
         0.03020333, -0.03673808], shape=(1536,), dtype=float32),
 array([-0.00053027,  0.00326079,  0.01927347, ..., -0.00887694,
         0.01718404, -0.02899886], shape=(1536,), dtype=float32),
 array([ 0.00783338, -0.02858108, -0.00320946, ...,  0.00291059,
         0.02553669, -0.01619502], shape=(1536,), dtype=float32),
 array([-0.00478341, -0.0259156 ,  0.05941101, ..., -0.00689318,
         0.00710032, -0.02335319], shape=(1536,), dtype=float32),
 array([-0.01993684, -0.02076978,  0.01989654, ...,  0.01121111,
         0.0137838

In [22]:
meta

[{'id': 0,
  'text': 'Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves'},
 {'id': 1,
  'text': 'himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves observing how machines work and is deeply fascinated by artificial intelligence and its impact on the IT industry.\n\nDeepak has strong technical skills in Python, Django, Flask, and other Python tools, along with knowledge of HTML, CSS, and JavaScript. He has worked on multiple projects including a Netflix homepage clone using HTML and CSS, an e-co

In [24]:
xb = np.vstack(embed_list)

In [25]:
xb

array([[ 0.02237804,  0.01795307, -0.00452745, ...,  0.01837507,
         0.03518271, -0.02058153],
       [-0.01348186, -0.0321741 ,  0.00224046, ..., -0.01961708,
         0.00827799, -0.01762411],
       [-0.03488477, -0.00471493, -0.02062185, ...,  0.02100492,
         0.0101577 , -0.03258635],
       ...,
       [-0.01993684, -0.02076978,  0.01989654, ...,  0.01121111,
         0.01378383, -0.00910189],
       [ 0.04160818, -0.02024985, -0.00312095, ...,  0.01302452,
         0.023751  ,  0.00761059],
       [ 0.0346989 , -0.01794612,  0.01301209, ...,  0.0166954 ,
         0.0255193 , -0.00946073]], shape=(10, 1536), dtype=float32)

In [46]:
xb.shape

(10, 1536)

In [26]:
import faiss

In [28]:
faiss.normalize_L2(xb)

# Why Do We Normalize Vectors Before Using FAISS?

Imagine you have vectors that represent sentences or words — these are just lists of numbers that show the “meaning” in mathematical form.

Before comparing how similar two vectors are (using Cosine Similarity), you need to make sure they are fairly compared — that’s where **L2 normalization** comes in.

---

## 🧠 Think of it like this:
Suppose two people are saying the same sentence:
- **Person A** says it loudly 🔊
- **Person B** says it softly 🤫

Both are saying the same thing, just with different volume.

- In math terms:
    - “Loudness” = **magnitude** (length) of the vector
    - “Meaning” = **direction** of the vector
- When we normalize, we make both voices the same loudness (set vector length = 1), so now we can focus only on meaning (the direction).

---

## 📏 What `faiss.normalize_L2(xb)` does:
It goes through every vector in your dataset (each sentence embedding) and divides it by its own length, so that:

$$
\text{new vector} = \frac{\text{original vector}}{|\text{original vector}|}
$$

After this step:
- Every vector’s length = 1
- Only the angle/direction matters (i.e., meaning)

---

## ⚙️ Why it’s important for FAISS:
FAISS searches for similar vectors very fast.

If you normalize the vectors first, then FAISS can simply use a **dot product** to find similar meanings — no need for long cosine calculations every time.

So this single line:
```python
faiss.normalize_L2(xb)
```
means:
> “Make all my sentence vectors the same length, so FAISS can compare their meanings directly and faster.”

---

## 🔢 Numerical Example (with 3D vectors)

### Step 1: Imagine two sentence embeddings
| Sentence              | Vector      |
|-----------------------|-------------|
| "Python is amazing"   | (2, 2, 1)   |
| "I love Python"       | (4, 4, 2)   |

These two sentences are similar in meaning, but the second one’s vector is just twice as large.

### Step 2: Find their lengths (L2 norm)
Formula:
$$
\text{Length} = \sqrt{x^2 + y^2 + z^2}
$$
- For (2,2,1):
  $$
  \sqrt{2^2 + 2^2 + 1^2} = \sqrt{9} = 3
  $$
- For (4,4,2):
  $$
  \sqrt{4^2 + 4^2 + 2^2} = \sqrt{36} = 6
  $$

### Step 3: Normalize them (make each vector’s length = 1)
We divide each component by its length.
- For (2,2,1):
  $$
  \left(\frac{2}{3}, \frac{2}{3}, \frac{1}{3}\right) = (0.67, 0.67, 0.33)
  $$
- For (4,4,2):
  $$
  \left(\frac{4}{6}, \frac{4}{6}, \frac{2}{6}\right) = (0.67, 0.67, 0.33)
  $$

✅ After normalization, both are exactly the same: **(0.67, 0.67, 0.33)**

### Step 4: What this means
- Before normalization:
    - (4,4,2) looked bigger — like a louder voice 🎤
    - (2,2,1) looked smaller — like a softer voice 🤫
    - But both were pointing in the same direction (same meaning).
- After normalization:
    - They both now have equal length (1)
    - FAISS can now compare only their direction, ignoring their size
    - So FAISS will correctly understand that both mean the same thing (because their direction in vector space is identical).

---

## 🧩 Why this matters
- If you **don’t normalize**:
    - FAISS may think (4,4,2) is more important just because it’s longer.
- If you **do normalize**:
    - FAISS focuses only on semantic similarity (the direction), not size.

---

## 🧠 Simple Analogy
Think of vectors like arrows 🎯:
- **Length** = how long the arrow is (volume or intensity)
- **Direction** = where it points (meaning)

L2 normalization simply shrinks or stretches all arrows so they’re the same length.
Now, when you compare them, you only see which way they point, not how long they are.

---

## 🎙️ Why FAISS gets confused without normalization
FAISS (and any search algorithm) measures similarity using dot product by default.

The dot product formula:
$$
\mathbf{a}\cdot\mathbf{b} = |\mathbf{a}| \times |\mathbf{b}| \times \cos(\theta)
$$
It depends on three things:
- Length of vector A
- Length of vector B
- Angle between them (θ = their direction difference)

So if one vector is longer, the dot product automatically becomes larger, even if both mean the same thing.

➡️ That’s why:
- (4,4,2) gives a bigger score than (2,2,1) — just because it’s longer.
- FAISS might mistakenly think it's more similar or more important, even though it’s not.

---

## ⚙️ What normalization fixes
When we normalize, we divide each vector by its length, making its size = 1.

| Vector      | Before         | After Normalization | Length |
|-------------|----------------|--------------------|--------|
| (2,2,1)     | (2,2,1)        | (0.67,0.67,0.33)   | 1      |
| (4,4,2)     | (4,4,2)        | (0.67,0.67,0.33)   | 1      |

✅ Both now have:
- Equal size (1)
- Same direction

So FAISS now compares only the direction, because all lengths are the same.

In the formula:
$$
\mathbf{a}\cdot\mathbf{b} = 1\times1\times\cos(\theta) = \cos(\theta)
$$
Now the dot product ignores size completely — it tells us only about angle (meaning).

---

## 🔍 Why this matters for “semantic” (meaning-based) search
Semantic search wants to find what’s similar in meaning, not what’s longer.

| Case                | What FAISS sees         | Result                        |
|---------------------|------------------------|-------------------------------|
| Without normalization | One vector is longer → looks “stronger” | Misleading result (not purely meaning-based) |
| With normalization    | All vectors same length → only direction matters | Correct result (true semantic similarity) |

---

## 🎯 Real-world analogy
Imagine two books:
- Book A: “Python is powerful.” (short book)
- Book B: “Python is powerful.” (same text, but with large font)

Both say the exact same thing, but the second book is physically bigger.
- Without normalization → FAISS compares the font size and thinks Book B is “more important.”
- With normalization → FAISS ignores the font size and focuses only on the words (meaning).

---

## 🧩 So when we say:
> “FAISS compares only direction, ignoring size.”

We mean:
- Every vector is first scaled to have the same length (1).
- Now, similarity depends only on how close their directions (angles) are in space.
- The closer the direction, the more similar the meaning.
- That’s why both (2,2,1) and (4,4,2) — after normalization — look identical to FAISS.

---

## 🧠 Your question (in simple words):
> “If FAISS is just a simple vector database that stores 512-dimensional embeddings, then why will it get confused without normalization? Aren’t the embeddings themselves already carrying the meaning?”

Perfect — that’s exactly the right thing to ask.

### ⚙️ What FAISS actually does
- FAISS does not understand meaning — it only performs mathematical vector search.
- It doesn’t know what “cat”, “dog”, or “AI” means.
- It only knows how to measure distance or similarity between numbers.

So if you give FAISS two vectors like:
- A = [4, 4, 2]
- B = [2, 2, 1]

it only sees them as points in space.

### 📏 How FAISS compares vectors
FAISS can use different similarity measures:
- L2 distance → measures straight-line (Euclidean) distance
- Dot product → measures angle and magnitude combined
- Cosine similarity → measures angle only (direction)

But FAISS doesn’t automatically know which you want.
So you must prepare the data correctly depending on which comparison you want.

### ⚖️ The role of normalization
Suppose you are using dot product search (most common when using normalized embeddings).
- Dot product depends on both:
    - Direction (meaning)
    - Magnitude (strength of the signal)
- If we don’t normalize, the magnitude (size of numbers) will influence the similarity score.
    - Vector [4,4,2] (large magnitude) gives a bigger dot product
    - Vector [2,2,1] (smaller magnitude) gives a smaller dot product
    - Even though they point in exactly the same direction (same meaning).
- Hence FAISS may think:
    > “Hmm, A looks more similar to the query than B”
    even though both mean the same thing!
- That’s what we mean by “FAISS gets confused.”

### 🎯 What normalization fixes
When we apply:
```python
faiss.normalize_L2(xb)
```
We scale every vector so its length = 1.
Now the dot product between two vectors = cosine of the angle between them.
That means:
- If two vectors point in the same direction → cosine = 1 (perfect match)
- If they point opposite → cosine = -1 (completely different)
- So now FAISS can focus purely on direction = meaning, and ignore the magnitude = loudness or intensity.

### 💡 Why this is essential for embeddings
- Embeddings from models like sentence-transformers or OpenAI embeddings already encode meaning — but they don’t guarantee the vector magnitudes are equal.
- So before using FAISS for semantic search, you normalize all embeddings once.
- That tells FAISS:
    > “Don’t worry about how long the vector is. Just compare how similar their meanings are.”

---

## 🧩 In short:
| Concept                | Without Normalization      | With Normalization   |
|------------------------|---------------------------|----------------------|
| FAISS compares         | Magnitude + Direction     | Direction only       |
| Meaning focus          | Lost or biased            | Preserved            |
| Similarity measure     | Dot Product (biased)      | Cosine Similarity    |
| Accuracy               | Lower                     | Higher               |

---

## ✨ You Don’t Need FAISS to Normalize Vectors!

Exactly, Deepak — you **don’t need FAISS** to normalize vectors. FAISS just provides a convenient function, but **normalization is pure math**. You can do it yourself using **NumPy** or any other library. ✅

Let me explain fully.

---

### 1️⃣ The normalization formula
For any vector $\mathbf{v} = [v_1, v_2, ..., v_n]$:
$$
\hat{\mathbf{v}} = \frac{\mathbf{v}}{|\mathbf{v}|}
$$
Where:
$$
|\mathbf{v}| = \sqrt{v_1^2 + v_2^2 + ... + v_n^2}
$$
- $|\mathbf{v}|$ = magnitude (length) of the vector
- $\hat{\mathbf{v}}$ = normalized vector (length = 1)

So you’re **dividing each component of the vector by its magnitude**.

---

### 2️⃣ How to do it with NumPy (example)
Suppose you have embeddings:
```python
import numpy as np

# Example 3D vectors
vectors = np.array([[2, 2, 1],
                    [4, 4, 2],
                    [2, 3, 1]])

# Compute L2 norm (magnitude) along axis=1 (rows)
norms = np.linalg.norm(vectors, axis=1, keepdims=True)

# Divide each vector by its magnitude
normalized_vectors = vectors / norms

print(normalized_vectors)
```
**Output:**
```
[[0.667, 0.667, 0.333],
 [0.667, 0.667, 0.333],
 [0.534, 0.801, 0.267]]
```
✅ Exactly the same as `faiss.normalize_L2(vectors)` does internally.

---

### 3️⃣ Important notes
- FAISS just provides a **fast built-in function** (`faiss.normalize_L2`) that works efficiently for **large datasets**, but the **math is exactly the same**.
- You can use **NumPy, PyTorch, or TensorFlow** — any library works.
- After normalization, you can even use **FAISS IndexFlatIP** or **your own similarity calculation** — it will behave the same.

---

### 4️⃣ Summary
| Method  | Formula                                                    | Works?                    |
| ------- | ---------------------------------------------------------- | ------------------------- |
| FAISS   | `faiss.normalize_L2(vectors)`                              | ✅ Yes, fast and optimized |
| NumPy   | `vectors / np.linalg.norm(vectors, axis=1, keepdims=True)` | ✅ Yes, same result        |
| PyTorch | `vectors / vectors.norm(p=2, dim=1, keepdim=True)`         | ✅ Yes, same result        |

---

💡 **Key insight:**
Normalization is **just math** — FAISS only provides a convenient function for large embeddings.
You can always do it manually.

---

In [33]:
dim = xb.shape[1]

In [34]:
dim

1536

# Understanding How FAISS Index Works: Step by Step

In [35]:
index = faiss.IndexFlatIP(dim)
index.add(xb)

No worries, let’s slow down and take this **step by step**, using a **simple analogy** so it clicks.

We are talking about these two lines:

```python
index = faiss.IndexFlatIP(dim)
index.add(xb)
```

I’ll explain in the **simplest possible way**.

---

## 1️⃣ Step 1: Creating the Index

### `index = faiss.IndexFlatIP(dim)`

Think of this like **creating an empty shelf in a library**.

- **`IndexFlatIP`** = a type of shelf that stores books (vectors) and can compare them by “content similarity.”
- **`dim`** = how big each book is (number of numbers in each vector, e.g., 512 dimensions).
- Right now, the shelf is **empty**. Nothing is stored yet.

**Analogy:**

| Code                             | Analogy                                                                                              |
| -------------------------------- | ---------------------------------------------------------------------------------------------------- |
| `index = faiss.IndexFlatIP(dim)` | “I’m creating an empty library shelf that can hold books of size `dim` and compare them by content.” |

---

## 2️⃣ Step 2: Adding Vectors to the Index

### `index.add(xb)`

Now you have a bunch of books (vectors) — `xb` is your collection.

- **`add(xb)`** = put all your books onto the shelf.
- FAISS now **remembers every book** so it can compare them later.

**Analogy:**

| Code            | Analogy                                                                                                                        |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------ |
| `index.add(xb)` | “I’m putting all my books on the shelf so that later, when someone asks for a book similar to a query, I can find it quickly.” |

---

## 3️⃣ Step 3: Why Inner Product (IP) Matters

- Inner product = dot product of two vectors.
- If you already normalized vectors (length = 1), **dot product = cosine similarity**.
- This means FAISS will now measure **how close the meanings are**, ignoring size.

**Analogy:**
- Each book has a theme (direction).
- FAISS checks: “Which books have the **same theme** as the query?”
- Because we normalized, it **doesn’t care about book thickness** (magnitude). Only theme (direction) matters.

---

## 4️⃣ Step 4: What Happens Internally

After `index.add(xb)`:

- FAISS stores all vectors in memory.
- Each vector has an **ID**.
- Later, when you give a query vector, FAISS will compare it to every stored vector using **dot product** and return the **most similar ones**.

**Analogy:**
- You hum a song → FAISS checks every song in your library → returns the **songs with the closest tune**.

---

## 🔑 Summary in Simplest Words

1. `index = faiss.IndexFlatIP(dim)` → “Create an empty library shelf that can compare books by meaning.”
2. `index.add(xb)` → “Put all my books (vectors) on the shelf so they are ready to be searched.”
3. FAISS will use **dot product** (cosine similarity if normalized) to find the closest matches.

---

In [36]:
index_path = "index_kumar.faiss"
meta_path = "meta_kumar.json"

In [37]:
faiss.write_index(index,index_path)

In [38]:
import json,os

In [39]:
with open(meta_path,'w') as f:
    for item in meta:
        f.write(json.dumps(item) + "\n")

In [47]:
query = "Who is deepak"
q= generate_embeddings(query).astype('float32').reshape(1,-1)
faiss.normalize_L2(q)
index.search(q,3)

(array([[0.65810657, 0.626843  , 0.56184447]], dtype=float32),
 array([[9, 8, 0]]))

# Why .reshape(1, -1) is Used in FAISS"

### Imagine a simple 1D array

```python
import numpy as np

arr = np.array([10, 20, 30, 40, 50, 60])
print(arr.shape)  # (6,)
```

Here, `arr` is a **1D vector** with 6 elements.

---

### Step 1: Reshape into 1 row

```python
new_arr = arr.reshape(1, -1)
print(new_arr.shape)  # (1, 6)
```

* `1` → We want **1 row** (like 1 query).
* `-1` → NumPy automatically calculates **6 columns** because all elements must fit.

So `reshape(1, -1)` transforms a 1D vector into a **2D matrix with 1 row**.

---

### Step 2: Reshape into 2 rows

```python
new_arr2 = arr.reshape(2, -1)
print(new_arr2.shape)  # (2, 3)
```

* `2` → We want **2 rows**.
* `-1` → NumPy calculates columns automatically → 3 columns per row.

**Result:**

```
[[10, 20, 30],
 [40, 50, 60]]
```

---

### ✅ Analogy

Think of it as **arranging marbles in rows and columns**:

* Original: a single line of 6 marbles.
* `reshape(1, -1)`: 1 row, 6 marbles in that row.
* `reshape(2, -1)`: 2 rows, NumPy figures out 3 marbles per row automatically.

---

In Faiss, `.reshape(1, -1)` is just like putting your **single query vector** into the right “shape” so the library can process it correctly.

---

# Testing whether after Normalization, the magnitude is approx 1 or not

In [48]:
test = np.array([[4,5,6,7,8]], dtype='float32')

In [49]:
np.linalg.norm(test)

np.float32(13.784049)

In [50]:
faiss.normalize_L2(test)

In [51]:
np.linalg.norm(test)

np.float32(0.99999994)

I am **testing and demonstrating the effect of L2 normalization** on a NumPy array, specifically by comparing the array's L2 norm **before** and **after** applying the Faiss L2 normalization function.

The image shows a three-step process:

***

## 1. Initial State: Creating the Array and Measuring its Norm

| Code | Purpose | Result |
| :--- | :--- | :--- |
| `test = np.array([[4,5,6,7,8]], dtype='float32')` | Creates a **2D NumPy array** containing a single vector (or row) of data. | `test` is the vector $[4, 5, 6, 7, 8]$. |
| `np.linalg.norm(test)` | Calculates the **L2 norm** (Euclidean length) of the vector. | `13.784049` (This is $\sqrt{4^2 + 5^2 + 6^2 + 7^2 + 8^2}$). |

This step establishes the initial, un-normalized L2 norm of the vector, which is approximately $13.78$.

***

## 2. The Normalization Step

| Code | Purpose |
| :--- | :--- |
| `faiss.normalize_l2(test)` | **Normalizes** the vector in-place. L2 normalization divides every element in the vector by the vector's L2 norm. |

The Faiss function modifies the `test` array so that its new L2 norm will be $\mathbf{1.0}$. Since this is an **in-place** operation, it changes the values stored within the original `test` variable.

***

## 3. Final State: Measuring the New Norm

| Code | Purpose | Result |
| :--- | :--- | :--- |
| `np.linalg.norm(test)` | Recalculates the L2 norm of the **modified** `test` array. | `1.0000000` (or a number very close to 1.0 due to floating-point precision). |

The final step **confirms that the `faiss.normalize_l2()` function worked correctly** by showing that the L2 norm of the vector has been successfully reduced (normalized) to **1.0**. This is the key characteristic of an L2-normalized vector.

# Difference Between `np.linalg.norm` and `faiss.normalize_L2`: Step-by-Step Guide

## 1️⃣ What Does `np.linalg.norm` Do?

- **Purpose:** Computes the **magnitude (length)** of a vector.
- **Does it change the vector?** No, it just gives a number.
- **Formula:** For a vector $v = [v_1, v_2, \dots, v_n]$:

$$
|v|_2 = \sqrt{v_1^2 + v_2^2 + \dots + v_n^2}
$$

- **Example:**

```python
import numpy as np

test = np.array([3, 4])
norm_val = np.linalg.norm(test)
print(norm_val)  # 5.0
```

Explanation: $\sqrt{3^2 + 4^2} = \sqrt{9 + 16} = \sqrt{25} = 5$ ✅

Here, `np.linalg.norm` **does not change `test`**. It just tells you its length.

---

## 2️⃣ What Does `faiss.normalize_L2` Do?

- **Purpose:** **Normalizes the vector** so its length becomes **1**. This is called **L2 normalization**.
- **Does it change the vector?** Yes, it **modifies the vector** in-place.
- **Formula:**

$$
v_{\text{normalized}} = \frac{v}{|v|_2}
$$

- **Example:**

```python
import faiss
import numpy as np

test = np.array([[3., 4.]])
faiss.normalize_L2(test)
print(test)
```

**Output:**
```
[[0.6, 0.8]]
```

Explanation:
- Original vector length = 5 (as before)
- Divide each element by 5 → `[3/5, 4/5] = [0.6, 0.8]`
- New length = 1 ✅

---

## 3️⃣ Key Differences Table

| Feature             | `np.linalg.norm`                | `faiss.normalize_L2`                                                                                 |
| ------------------- | ------------------------------- | ---------------------------------------------------------------------------------------------------- |
| What it does        | Computes the length of a vector | Scales the vector to have length 1                                                                   |
| Output              | A **number**                    | A **modified vector**                                                                                |
| Changes the vector? | ❌ No                            | ✅ Yes (in-place)                                                                                     |
| Use case            | Just want magnitude             | Preprocessing vectors for similarity search (FAISS expects normalized vectors for cosine similarity) |

---

## 💡 Analogy

- `np.linalg.norm`: “**Measure the rope** to see how long it is.”
- `faiss.normalize_L2`: “**Stretch or shrink the rope** so it’s exactly 1 meter long.”

---

## Why Normalizing is Crucial in FAISS

In Faiss, **normalizing is crucial** because when using **cosine similarity**, FAISS actually computes it as a **dot product**, and vectors must be normalized for this to correctly represent similarity.

---

## 4️⃣ Visual Example: Norm vs Normalize

### Example: Using `np.linalg.norm`

```python
import numpy as np

test = np.array([3, 4])
val = np.linalg.norm(test)
print(val)
```

**Output:**
```
5.0
```

Here’s what happens step by step:

| Step | Description               | Result    |
| ---- | ------------------------- | --------- |
| 1    | Takes the vector `[3, 4]` | →         |
| 2    | Squares each element      | `[9, 16]` |
| 3    | Adds them up              | `25`      |
| 4    | Takes square root         | `√25 = 5` |

So the **magnitude (length)** of `[3, 4]` is `5`.
That’s all it does — just gives you **how long** the vector is.

---

### ⚠️ It Doesn’t Normalize

If you print `test` again:

```python
print(test)
```

You’ll still get:
```
[3 4]
```
So the vector itself is **unchanged**.

---

## 5️⃣ Summary Table: Norm vs Normalize

| Function                   | What It Does                                                     | Changes Vector? | Output                            |
| -------------------------- | ---------------------------------------------------------------- | --------------- | --------------------------------- |
| `np.linalg.norm(test)`     | Calculates **magnitude only**                                    | ❌ No            | A **number** (e.g., `5.0`)        |
| `faiss.normalize_L2(test)` | Divides each element by magnitude to make vector **unit length** | ✅ Yes           | A **vector** (e.g., `[0.6, 0.8]`) |

---

## 🧠 Analogy Recap

- `np.linalg.norm` → “How long is this arrow?” 🧮
- `faiss.normalize_L2` → “Make this arrow exactly **1 unit long**, pointing in the same direction.” 🎯

---

## 6️⃣ Why Do We Use `np.linalg.norm()` At All?

Because before you can **normalize** a vector (make its length = 1), you need to know **what its current length** is.
That’s what `np.linalg.norm()` gives you — the **length (magnitude)** of the vector.

---

## 7️⃣ How Normalization Works: Step-by-Step

Suppose your vector is:

```python
v = np.array([3, 4])
```

Now you want to make it a **unit vector** (length = 1).
How will you do that?
You first need to know how long it is right now!

### Step-by-step:

1️⃣ **Find its magnitude:**
```python
length = np.linalg.norm(v)
print(length)  # 5.0
```

2️⃣ **Divide each element by the length:**
```python
normalized = v / length
print(normalized)  # [0.6, 0.8]
```

Now the new vector `[0.6, 0.8]` has a **length of 1**.
✅ This process is called **L2 normalization**.

---

## 8️⃣ Why the Name `norm`?

Because `norm` is the **mathematical name** for the *length* of a vector — and it’s used as the **basis for normalization**.

> **Normalization = Divide by the norm.**

That’s why:
- `np.linalg.norm()` → calculates the **norm (length)**
- `faiss.normalize_L2()` → uses the **norm internally** to make every vector have length 1

---

## 9️⃣ What Does FAISS Do Internally?

When you call:
```python
faiss.normalize_L2(xb)
```
FAISS is doing this internally for each row vector:
```python
xb[i] = xb[i] / np.linalg.norm(xb[i])
```
So FAISS is using the *concept* of norm — it just automates it for you.

---

## 1️⃣0️⃣ Final Summary Table

| Function                | What It Does                    | Uses “norm”?           | Purpose                     |
| ----------------------- | ------------------------------- | ---------------------- | --------------------------- |
| `np.linalg.norm(x)`     | Finds the length of the vector  | ✅ Directly computes it | To measure magnitude        |
| `faiss.normalize_L2(x)` | Divides vector by its norm (L2) | ✅ Uses norm internally | To make vectors unit length |

---

**In short:**
We write or use `norm` because *it’s the foundation of normalization.*
Without knowing the length, we can’t make the vector length = 1. 💪

---

In [67]:
query = "what is the qualification of deepak"
q= generate_embeddings(query).astype('float32').reshape(1,-1)
faiss.normalize_L2(q)
search = index.search(q,3)
for i in search[1][0]:
    print(chunks[i])


erall, Deepak is a motivated and passionate learner with strong technical skills, a growing project portfolio, and a clear vision to become a successful Data Scientist.
Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves
ld AI-powered applications.

Deepak follows Vishwa Mohan on YouTube for inspiration and learning. His personality blends technical curiosity with dedication and creativity. He enjoys exploring new concepts, sharing insights with others, and documenting his journey in tech. He values clarity and precision in communication and pre

In [63]:
search

(array([[0.6580955 , 0.62682927, 0.5618104 ]], dtype=float32),
 array([[9, 8, 0]]))

In [68]:
print(search[1][0])

[9 0 8]


In [66]:
for i in search[1][0]:
    print(chunks[i])

erall, Deepak is a motivated and passionate learner with strong technical skills, a growing project portfolio, and a clear vision to become a successful Data Scientist.
ld AI-powered applications.

Deepak follows Vishwa Mohan on YouTube for inspiration and learning. His personality blends technical curiosity with dedication and creativity. He enjoys exploring new concepts, sharing insights with others, and documenting his journey in tech. He values clarity and precision in communication and prefers avoiding pronouns in datasets to keep information unambiguous. Overall, Deepak is a motivated and passionate learner with strong technical skills, a growing project 
Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scient

In [54]:
chunks

['Deepak Kumar Mohanty is from Balasore, Odisha, India. He completed his BCA at Bhadrak Autonomous College, Odisha, and is currently preparing for a Data Scientist role in the IT industry. Deepak is 21 years old, 172 cm tall, and weighs 91 kg. He is highly ambitious, committed to proving that it is possible to secure a Data Scientist job without a Master’s degree within 6 to 12 months. He describes himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves',
 'himself as a fast learner, a good listener, and very flexible in adopting new technologies. He loves observing how machines work and is deeply fascinated by artificial intelligence and its impact on the IT industry.\n\nDeepak has strong technical skills in Python, Django, Flask, and other Python tools, along with knowledge of HTML, CSS, and JavaScript. He has worked on multiple projects including a Netflix homepage clone using HTML and CSS, an e-commerce website clone, and a React calcula